aptos-move/framework/aptos-framework/sources/AuctionHouse.move

```diff
@@ -0,0 +1,878 @@
+    ) acquires CoinEscrow, ListingData {
+        let sender_addr = Signer::address_of(sender);
+        assert!(sender_addr != seller, EINVALID_BUYER);
+
+        let listing_id = GUID::create_id(seller, guid_creation_num);
+        let listing_data = borrow_global_mut<ListingData>(seller);
+        let listed_items = &mut listing_data.listed_items;
+        assert!(Table::contains(listed_items, listing_id), ELISTING_INACTIVE);
+        let listed_item = Table::borrow_mut(listed_items, listing_id);
+        assert!(is_auction_active(listed_item.start_time, listed_item.duration), ELISTING_INACTIVE);
+
+        let current_bid = Option::get_with_default(&listed_item.current_bid, 0);
+        if (current_bid != 0) { // bids should increase by 5% from the previous bids 
+            assert!(bid * 100 > (current_bid * 105), EINSUFFICIENT_BID);
+        } else { // if this the first bid, then it can start from min_selling_price
+            assert!(bid > listed_item.min_selling_price-1, EINSUFFICIENT_BID);
+        };
+
+        // send back the deposit of the current bidder only if someone has bid before
+        let current_bidder = Option::get_with_default(&listed_item.current_bidder, seller);
+        if (current_bidder != seller) {
+            let current_bidder_locked_coins = &mut borrow_global_mut<CoinEscrow>(current_bidder).locked_coins;
+            let coins = Table::remove(current_bidder_locked_coins, listing_id);
+            Coin::deposit<TestCoin>(current_bidder, coins);
+        };
+
+        if (Option::get_with_default(&listed_item.buy_now_price, 18446744073709551615) <= bid) { // max integer
+            return buy_now(sender, seller, listing_id, bid)
+        };
+
+        Event::emit_event<BidEvent>(
+            &mut listing_data.bid_events,
+            BidEvent { id: listing_id, bid: bid },
+        );
+
+        if (!exists<CoinEscrow>(sender_addr)) {
+            move_to(sender, CoinEscrow {
+                locked_coins: Table::new<ID, Coin<TestCoin>>()
+            });
+        };
+
+        let locked_coins = &mut borrow_global_mut<CoinEscrow>(sender_addr).locked_coins;
+        let coins = Coin::withdraw<TestCoin>(sender, bid);
+        Table::add(locked_coins, listing_id, coins);
+
+        if (current_bidder == seller) {
+            Option::fill(&mut listed_item.current_bidder, sender_addr);
+            Option::fill(&mut listed_item.current_bid, bid);
+        } else {
+            Option::swap(&mut listed_item.current_bidder, sender_addr);
+            Option::swap(&mut listed_item.current_bid, bid);
+        };
+    }
+
+    // claims the token from the seller's account
+    // called by the winning bidder, if no one bid on the auction then, winning bidder = seller
+    // if the seller has not claimed their coins, deposit the escrowed coins to the seller's account
+    // if both claims have been made, remove the auction item from the seller's auction data
+    public(script) fun claim_token_script(
```

Comment:
I think this could be saved if we do token and coin transfer directly in the bid function

Reply:
bid can only claim and transfer coins if it is a buy bid